usr/src/lib/libnsl/rpc/svc_generic.c

```diff
@@ -106,6 +107,221 @@ __svc_free_xprtlist(void)
+ */
+int
+svc_create_port(void (*dispatch)(), const rpcprog_t prognum,
+    const rpcvers_t versnum, const char *nettype, in_port_t port)
+{
+	void *handle;
+	struct netconfig *nconf, *tcp6_nconf = NULL;
+	struct t_info tinfo;
+	struct t_bind *bind_addr, *reqb;
+	int fd, sv_errno;
+	boolean_t tcp_done = B_FALSE;
+	char error_buf[256];
+	int num_reg = 0;
+	SVCXPRT_LIST *l;
+
+	if (ntohs(port) == 0) {
+		return (svc_create(dispatch, prognum, versnum, nettype));
+	}
+
+	handle = __rpc_setconf((char *) nettype);
+	if (handle == NULL) {
+		syslog(LOG_ERR, "Failed to retrieve rpc conf handle for %s\n",
+		    nettype);
+		return (0);
+	}
+
+	/*
+	 * We loop through every device in the given nettype, registering the
+	 * service on that device.  However, there is a small catch. When we
+	 * register the service on the tcp6 device, it registers for both ipv4
+	 * and ipv6, preventing us from registering through the tcp device.
+	 * This causes many machines to not be able to connect to the service
+	 * we're registering, even though an ipv4 address is being presented
+	 * through tcp6. The simplest way around this is to register tcp before
+	 * tcp6; this loop does that. When tcp first comes up, it sets a flag so
+	 * that we know we've completed tcp.  When tcp6 comes up in the list,
+	 * there are two options.  If we've already completed tcp, then we
+	 * proceed with tcp6.  If not, then we store it in tcp6_nconf and
+	 * continue until we find tcp.  After we finish tcp, the next time
+	 * through the loop, we see that tcp_done is set and tcp6_nconf is set,
+	 * so we execute that, rather than pulling another device off the rpc
+	 * handle.
+	 */
+	while ((tcp_done && (nconf = tcp6_nconf) != NULL) ||
+	    (nconf = (struct netconfig *)__rpc_getconf(handle))) {
+		SVCXPRT *xprt;
+
+		if (strcmp(nconf->nc_netid, "tcp") == 0) {
+			tcp_done = B_TRUE;
+		} else if (strcmp(nconf->nc_netid, "tcp6") == 0) {
+			if (tcp_done) {
+				tcp6_nconf = NULL;
+			} else {
+				tcp6_nconf = nconf;
+				continue;
+			}
+		}
+
+		/*
+		 * We loop through the services we've already registered.
+		 * If we find a service already registered for this device,
+		 * we add our version number to it.  If we were to instead
+		 * create a new service, some of the versions might be
+		 * inaccessible.
+		 */
+		for (l = _svc_xprtlist; l; l = l->next) {
+			if (strcmp(l->xprt->xp_netid, nconf->nc_netid) == 0) {
+				/* Found an old one, use it */
+				(void) rpcb_unset(prognum, versnum, nconf);
+				if (svc_reg(l->xprt, prognum, versnum,
+				    dispatch, nconf) == FALSE)
+					(void) syslog(LOG_ERR,
+					    "svc_create_port: could not "
+					    "register prog %d vers %d on %s",
+					    prognum, versnum, nconf->nc_netid);
+				else
+					num_reg++;
+				break;
+			}
+		}
+		(void) mutex_unlock(&xprtlist_lock);
+		if (l != NULL)
+			continue;
+
+		/*
+		 * Intentionally leaked, since closing it terminates the
```

Comment:
Thanks, @mtelka .  In that case, the comment is in error; the fd isn't leaked, because we end up calling svc_tli_create, which will store it in the xprt.  When the service is destroyed, we will close and free the fd.  The code should be tweaked to not free or destroy the fd in cases where we also destroy xprt, though.